src/main/java/graphql/parser/UnicodeUtil.java

```diff
@@ -0,0 +1,64 @@
+package graphql.parser;
+
+import graphql.Assert;
+import graphql.Internal;
+
+import java.io.IOException;
+import java.io.StringWriter;
+
+/**
+ * Contains Unicode helpers for parsing StringValue types in the grammar
+ */
+@Internal
+public class UnicodeUtil {
+    public static int MAX_UNICODE_CODE_POINT = 0x10FFFF;
+
+    public static int parseAndWriteUnicode(StringWriter writer, String string, int i) {
+        // Unicode characters can either be:
+        //  - four hex characters in the form \\u597D, or
+        //  - any number of hex characters surrounded by a brace in the form \\u{1F37A}
+
+        // Four hex character only case \\u597D, for code points in the Basic Multilingual Plane (BMP)
+        if (isNotBracedEscape(string, i)) {
+            String hexStr = string.substring(i + 1, i + 5);
+            int codepoint = Integer.parseInt(hexStr, 16);
+            writer.write(codepoint);
+            return i + 4;
+            // TODO error checking of invalid values
+        } else {
+            // Any number of hex characters e.g. \\u{1F37A}, which allows code points outside the Basic Multilingual Plane (BMP)
+            int startIx = i + 2;
+            int endIndexExclusive = startIx;
+            do {
+                if (endIndexExclusive + 1 >= string.length()) {
+                    throw new RuntimeException("invalid unicode encoding");
+                }
+            } while (string.charAt(++endIndexExclusive) != '}');
+
+            String hexStr = string.substring(startIx, endIndexExclusive);
+            Integer hexValue = Integer.parseInt(hexStr, 16);
+            if (isValidUnicodeCodePoint(hexValue)) {
+                char[] chars = Character.toChars(hexValue);
+                try {
+                    writer.write(chars);
+                } catch (IOException e) {
+                    return Assert.assertShouldNeverHappen();
+                }
+                return endIndexExclusive;
+            } else {
+                throw new RuntimeException("invalid unicode code point");
```

Comment:
Put the offening value in the exception - make this a GraphqlException

Reply:
Or will this get re-translated by ANTLR into something else

We want  a graphql.parser.InvalidSyntaxException at the end

Reply:
There are two sides of this Unicode RFC:
1. Unescaped unicode characters like 🍺
2. Escaped unicode characters such as `\u{1F37A}` (escaped version of 🍺)

Unescaped unicode characters are parsed by ANTLR and will raise the usual `InvalidSyntaxException`.

Escaped unicode characters are parsed separately by this `UnicodeUtil`, which is called by `parseSingleQuotedString()`. As this is outside ANTLR, exceptions are not automatically re-translated - I'll add this.